itest/assertions.go

```diff
@@ -92,55 +94,90 @@ func assetScriptKeyIsLocalCheck(isLocal bool) assetCheck {
 	}
 }
 
+// sortAssetsByName converts an unordered list of assets to a map of lists of
+// assets, where all assets in a list have the same name.
+func sortAssetsByName(assets []*taprpc.Asset) map[string][]*taprpc.Asset {
+	assetLists := make(map[string][]*taprpc.Asset)
+	for _, a := range assets {
+		if _, ok := assetLists[a.AssetGenesis.Name]; !ok {
+			assetLists[a.AssetGenesis.Name] = []*taprpc.Asset{}
+		}
+
+		assetLists[a.AssetGenesis.Name] = append(
+			assetLists[a.AssetGenesis.Name], a,
+		)
+	}
+
+	return assetLists
+}
+
 // assertAssetState makes sure that an asset with the given (possibly
 // non-unique!) name exists in the list of assets and then performs the given
 // additional checks on that asset.
-func assertAssetState(t *harnessTest, tapd *tapdHarness, name string,
-	metaHash []byte, assetChecks ...assetCheck) *taprpc.Asset {
+func assertAssetState(t *harnessTest, assets map[string][]*taprpc.Asset,
+	name string, metaHash []byte, assetChecks ...assetCheck) *taprpc.Asset {
 
-	t.t.Helper()
+	var a *taprpc.Asset
+	matchingAssets, ok := assets[name]
+	require.True(
+		t.t, ok, fmt.Errorf("no assets named %s in asset list", name),
+	)
 
-	ctxb := context.Background()
+	for _, rpcAsset := range matchingAssets {
+		rpcGen := rpcAsset.AssetGenesis
+		if bytes.Equal(rpcGen.MetaHash, metaHash[:]) {
+			a = rpcAsset
 
-	var a *taprpc.Asset
-	err := wait.NoError(func() error {
-		ctxt, cancel := context.WithTimeout(ctxb, defaultWaitTimeout)
-		defer cancel()
+			for _, check := range assetChecks {
+				err := check(rpcAsset)
+				require.NoError(t.t, err)
+			}
 
-		listResp, err := tapd.ListAssets(
-			ctxt, &taprpc.ListAssetRequest{},
-		)
-		if err != nil {
-			return err
+			break
 		}
+	}
 
-		for _, rpcAsset := range listResp.Assets {
-			rpcGen := rpcAsset.AssetGenesis
-			if rpcGen.Name == name &&
-				bytes.Equal(rpcGen.MetaHash, metaHash[:]) {
+	require.NotNil(t.t, a, fmt.Errorf("asset with matching metadata not"+
+		"found in asset list"))
 
-				a = rpcAsset
+	return a
+}
 
-				for _, check := range assetChecks {
-					if err := check(rpcAsset); err != nil {
-						return err
-					}
-				}
+// waitForBatchState polls until the planter has reached the desired state with
+// the current batch.
+func waitForBatchState(t *harnessTest, ctx context.Context, tapd *tapdHarness,
+	timeout time.Duration, targetState mintrpc.BatchState) bool {
 
-				break
-			}
-		}
+	breakTimeout := time.After(timeout)
+	ticker := time.NewTicker(50 * time.Millisecond)
```

Comment:
Should we extract all of this into a copy of `wait.NoError` and `wait.Predicate` analogues that allow us to specify the poll interval instead? In the `fn` package for example?

Reply:
Yes, the intent was to separate waiting for the minter to reach a certain state, and asserting properties over the batch.

The slowdown before was caused by two features of the previous code:

- Using `wait.NoError` to actually wait for the planter state, meant we also used a `wait.NoError` call in every call of `assertAssetState`.
- `assertAssetState` called `ListAssets()`, so we would also call that RPC for each entry in the batch.

So, for asserting properties over larger batches (even just 100 assets), this caused the whole test to run much slower.

The intent now is to wait for the desired batch state, call `ListAssets()` once, and then run the assertions over the results without any waiting logic or extra RPC calls.

`ListAssets()` actually still slows down as batch size increases, but that's sort of a separate issue. `waitForBatchState` is very similar to `waitForNTXsInMempool`, I agree that this pattern could be moved into `fn`.